design/src/rfcs/rfc0041_identity_cache_partitions.md

```diff
@@ -0,0 +1,293 @@
+different clients.
+
+### Design intent
+
+Identity providers and identity caching are intentionally decoupled. This allows caching behavior to be more easily
+customized and centrally configured while also removing the need for each identity provider to have to implement
+caching. There is some fallout from sharing an identity cache though. This is fairly well documented on 
+`IdentityCachePartition` itself. 
+
+```rust,ignore
+/// ...
+///
+/// Identities need cache partitioning because a single identity cache is used across
+/// multiple identity providers across multiple auth schemes. In addition, a single auth scheme
+/// may have many different identity providers due to operation-level config overrides.
+///
+/// ...
+pub struct IdentityCachePartition(...)
+```
+
+Cache partitioning allows for different identity types to be stored in the same cache instance as long as they
+are assigned to a different partition. Partitioning also solves the issue of overriding configuration on a per operation
+basis where it would not be the correct or desired behavior to re-use or overwrite the cache if a different resolver
+is used. 
+
+In other words cache partitioning is effectively tied to a particular instance of an identity resolver. Re-using the
+same instance of a resolver _SHOULD_ be allowed to share a cache partition. The fact that this isn't the case
+today is an oversight in how types are wrapped and threaded through the SDK.
+
+
+The user experience if this RFC is implemented
+----------------------------------------------
+
+In the current version of the SDK, users are unable to share cached results of identity resolvers via shared `SdkConfig`
+across clients.
+
+Once this RFC is implemented, users that create clients via `SdkConfig` with the latest behavior version will share
+a default identity cache. Shared identity resolvers (e.g. `credentials_provider`, `token_provider`, etc) will provide 
+their own cache partition that is re-used instead of creating a new one each time a provider is converted into a 
+`SharedIdentityResolver`.
+
+### Default behavior
+
+```rust,ignore
+let config = aws_config::defaults(BehaviorVersion::latest())
+    .load()
+    .await;
+
+let c1 = aws_sdk_foo::Client::new(&config);
+c1.foo_operation().send().await;
+
+
+let c2 = aws_sdk_bar::Client::new(&config);
+// will re-use credentials/identity resolved via c1
+c2.bar_operation().send().await;
+```
+
+Operations invoked on `c2` will see the results of cached identities resolved by client `c1` (for operations that use
+the same identity resolvers). The creation of a default identity cache in `SdkConfig` if not provided will be added
+behind a new behavior version.
+
+### Opting out
+
+Users can disable the shared identity cache by explicitly setting it to `None`. This will result in each client
+creating their own identity cache.
+
+```rust,ignore
+let config = aws_config::defaults(BehaviorVersion::latest())
+    .set_identity_cache(None)
```

Comment:
This apparently doesn't exist on `ConfigLoader`, it only exists on the explicit client config. We'll probably instead need something like `no_identity_cache()` to match `no_credentials()`